File: server/src/main/java/io/microconfig/server/vault/VaultKVSecretResolverStrategy.java
package io.microconfig.server.vault;

import io.microconfig.core.Microconfig;
import io.microconfig.core.properties.DeclaringComponentImpl;
import io.microconfig.core.properties.Placeholder;
import io.microconfig.core.properties.PlaceholderResolveStrategy;
import io.microconfig.core.properties.Property;
import io.microconfig.server.configs.DynamicVarsResolverStrategy;
import lombok.RequiredArgsConstructor;

import java.util.HashMap;
import java.util.Optional;

import static io.microconfig.core.configtypes.ConfigTypeFilters.configTypeWithName;
import static io.microconfig.core.properties.ConfigFormat.PROPERTIES;
import static io.microconfig.core.properties.PropertyImpl.property;
import static io.microconfig.server.vault.VaultConfig.vaultConfig;
import static java.util.Optional.empty;
import static java.util.Optional.of;

@RequiredArgsConstructor
public class VaultKVSecretResolverStrategy implements PlaceholderResolveStrategy {
    private final Microconfig microconfig;
    private final DynamicVarsResolverStrategy dynamicVars;
    private VaultConfig vaultConfig;
    private VaultClient vaultClient;

    @Override
    public Optional<Property> resolve(Placeholder placeholder) {
        if (!"VAULT-KV".equals(placeholder.getComponent())) return empty();
        if (vaultConfig == null) {
            var properties = microconfig.inEnvironment(placeholder.getEnvironment())
                .getComponentWithName(placeholder.getRootComponent())
                .getPropertiesFor(configTypeWithName(placeholder.getConfigType()))
                .withPrefix("microconfig.vault")
                .resolveBy(microconfig.resolver())
                .getPropertiesAsKeyValue();

            var vars = new HashMap<>(properties);
            vars.putAll(dynamicVars.dynamicVars());

            vaultConfig = vaultConfig(vars);
            vaultClient = new VaultClientImpl(vaultConfig);
        }

        String secret = vaultClient.fetchKV(placeholder.getKey());
        return of(property(placeholder.getKey(), secret, PROPERTIES,
                new DeclaringComponentImpl(placeholder.getConfigType(), "HashiCorp Vault", placeholder.getEnvironment())));
    }
}

File: server/src/main/kotlin/io/microconfig/server/vault/VaultKVSecretResolverStrategy.kt
package io.microconfig.server.vault

import io.microconfig.core.Microconfig
import io.microconfig.core.configtypes.ConfigTypeFilters.configTypeWithName
import io.microconfig.core.properties.ConfigFormat.PROPERTIES
import io.microconfig.core.properties.DeclaringComponentImpl
import io.microconfig.core.properties.PlaceholderResolveStrategy
import io.microconfig.core.properties.Property
import io.microconfig.core.properties.PropertyImpl.property
import io.microconfig.server.common.toOptional
import java.util.*
import java.util.Optional.empty

class VaultKVSecretResolverStrategy(val microconfig: Microconfig) : PlaceholderResolveStrategy {

    private var vaultClient: VaultClient? = null

    override fun resolve(
        component: String,
        key: String,
        environment: String,
        configType: String,
        root: String
    ): Optional<Property> {
        if ("VAULT-KV" != component) return empty()
        val secret = client(environment, root, configType).fetchKV(key)

        val parent = DeclaringComponentImpl(configType, "HashiCorp Vault", environment)
        return property(key, secret, PROPERTIES, parent).toOptional()
    }

    private fun client(environment: String, root: String, configType: String): VaultClient {
        vaultClient = vaultClient ?: initClient(environment, root, configType)
        return vaultClient!!
    }

    private fun initClient(environment: String, root: String, configType: String): VaultClient {
        val properties = microconfig.inEnvironment(environment)
            .getComponentWithName(root)
            .getPropertiesFor(configTypeWithName(configType))
            .withPrefix("microconfig.vault")
            .resolveBy(microconfig.resolver())
            .propertiesAsKeyValue

        val vaultConfig = vaultConfig(properties)
        return VaultClientImpl(vaultConfig)
    }
}